### PR TITLE
[WIP] Dockerization

### PR DESCRIPTION
**Pull request by inakoll**
I've started to implement the docker back-end for kevin and I've also added a Dockerfile to build an image that should be able to run kevin, falk or chantal. In the end, chantal should use a user provided Dockerfile while kevin and falk could keep this one or fork it if necessary.
- [x] Add docker back-end
- [x] Add a Dockerfile for kevin
- [x] Fix Python version in doc/setup.md
- [x] Figuring out how to launch kevin, falk and chantal (using systemd inside the container)
- [ ] Develop an SSH agnostic Falk protocol so that Kevin talks to Falk instead of ssh-ing inside the falk-provided machine
- [ ] Test kevin, falk and chantal docker containers on the same box

Any help appreciated : )

### CODE REVIEW

**Comment by TheJJ** [comment]
hmh i'd avoid adding a dockerfile for kevin itself, and if we want to ship one, we should base it on a more recent version of some distro, maybe debian testing, maybe arch, but ubuntu 14.04 is two years old :)

You can lauch kevin with `python -m kevin`, falk with `python -m falk`, `-c` can specify the config location. `--help` will of course help you.

File: falk/vm/docker.py [hunk 1]
```diff
@@ -3,3 +3,59 @@
...
+        super().__init__(cfg)
+        self.manage = False
+        self.docker_container_id = None
+
+    @classmethod
+    def config(cls, machine_id, cfgdata, cfgpath):
+        cfg = ContainerConfig(machine_id, cfgdata, cfgpath)
+
+        cfg.dockerfile = Path(cfgdata["dockerfile"])
+        cfg.image_name = Path(cfgdata["image_name"])
```
Comment (TheJJ) [inline falk/vm/docker.py]: If the path is relative, you should make it relative to `cfgpath` so it remains intuitive to use.

**Comment by TheJJ** [comment]
Still interested in adding a docker backend? Then we could run falk in a VM and spawn docker containers there :)

**Comment by iNaKoll** [comment]
Yep still there! Still interested! I just have to find some time to work on this. I can't promise anything but I might have some time for this this weekend

**Comment by iNaKoll** [comment]
ping @TheJJ please, take another look!
This is still a work in progress but I've completely reworked this most expected falk vm provider.
I've just unit tested this provider (python -m falk.vm.docker) and it mostly work.
If you have some time, I might need some help to finish the integration with the rest of kevin.

- [x] implement docker container provisioning
- [x] add a Container.execute method that doesn't rely on ssh
- [] test and debug docker.download/upload
- [] test and debug falk/chantal/kevin integration

**Comment by iNaKoll** [comment]
I've just pushed some fixes for Docker.execute/upload/download and updated the unit test.
This docker provider should just work now.
There is still some integration tests to be done with the rest of kevin.
Just let me know if you think you could help me with that.
Thanks

**Comment by TheJJ** [comment]
Hai! that approach looks cool but also quite complex, to represent in-docker processes directly as Popen-processes :) I'm a bit busy the nest days but I'll come back to you so we can ship this.

Do you have specific questions on how to integrate this with the rest of kevin?

~~Another idea I've had (didn't see it in your patch yet) is to design the docker backend in such a way that one has a separate git repo that defines the dockerfile and funny additional files. Pull requests to this repo also submit a webhook to kevin, but not to test it, but to update the container configuration instead. That way, we don't need to clone the repo on the host~~
Oh wait. Now that I think about it: We can clone the repo on the host and ship a dockerfile. This dockerfile defines the container setup (if nothing is changed, an already built container can be used) and the container is rebuilt when needed. The cloned repo is then mounted into the container for the build. What do you think?

**Comment by iNaKoll** [comment]
> Hai! that approach looks cool but also quite complex, to represent in-docker processes directly as Popen-processes :) I'm a bit busy the nest days but I'll come back to you so we can ship this.

The advantage of this Popen-like approach is that I don't have to reimplement a DockerProcessTransport class from scratch. I've found that it's easier to reimplement a DockerPopen-like process class and use BaseSubprocessTransport as a base class for _DockerProcessTransport. The downside is that I have to use an os.pipe instead of queue.Queue and/or asyncio.Queue for the process input and output streams. I am all ears if you have any suggestion that would simplify this module.

> Do you have specific questions on how to integrate this with the rest of kevin?

I don't really have any specific question on how to integrate this with the rest of kevin. I think that, _in time_, there is nothing I wouldn't be able to figure out by myself for this PR.
However, please note that for now : 
- I have only unit-tested this vm provider (python -m falk.vm.docker)
- Most of the files diff outside falk/vm/* have not been worked-on for _years_.
- I didn't have the opportunity to check that everything was working properly (i.e. run some CI jobs with both QEMU and Docker) => There are some bugs waiting around the corner here.

Especially, I don't have any idea of why 2 years ago I've made some changes in: 
- kevin/simulator/github.py
- kevin/util.py
- kevin/process.py:290
- falk/shell.py
- falk/__main__.py
- falk/__init__.py

So if you have some time to test this PR, run some CI jobs and give me some feedback, that would really be helpful.

> Oh wait. Now that I think about it: We can clone the repo on the host and ship a dockerfile. This dockerfile defines the container setup (if nothing is changed, an already built container can be used) and the container is rebuilt when needed. The cloned repo is then mounted into the container for the build. What do you think?

Yes, I think that's a good idea. Maybe we could just use `git archive` to download the .kevin file and then if we're asked to provision a docker container download the .kevin.d/docker/ directory that would contain the Dockerfile and the docker build context. That way we wouldn't have to clone the whole git repo to build/rebuild a container image.

**Comment by iNaKoll** [comment]
Ok, I've done  some digging, fixed some bugs and set up a docker test environment for Kevin (see Dockerfile and docker/ directory).

I think that the main remaining integration issue is that the FalkVM class in kevin is currently assuming that it can directly talk to the container over ssh without relying on the falk process as an intermediate (see the wait_for_ssh function).

The docker falk provider does not rely on an ssh server running inside the container. So I think that kevin should not try to talk to the container directly. It just make thing overcomplicated.

I currently can't wrap my head around all the interfaces involved here. I'll have to come up with some ideas eventually.

**Comment by TheJJ** [comment]
Ah yes, that is not ideally designed, currently. There needs to be an indirection of how the container is actually controlled, the `ssh`-control is only one of the backends then. Maybe other falk setups are controllable by some funny http API. 

We need some support for streaming the output and an independent control channel. With `ssh` we use `stdout` and `stderr` for the two channels. For controlling docker, we could relay the connection through falk and just tag the data with `control` or `output`, as falk then owns the docker container processes anyway.

**Comment by TheJJ** [comment]
We now (#40) have support for podman containers as falk backend, so I'll close this for now. If we wanted to add docker some day, we should do it very similar to the podman implementation.

**Pull request closed without merging**